### PR TITLE
Canvas middleware transfer

### PR DESCRIPTION
**Pull request by MaximilianMeiler**
Added the following changes:

- Incorporated all canvas calls in the middleware's "login" route
- Redid course data structure to split upcoming assignments and submitted assignments
- Added canvas route for efficient submission collection
- Adjusted frontend display to handle this new data structure
- Removed logout route, set loginTime directly through middleware on login
- Removed newly unused states/effects from app.js

### CODE REVIEW

**Comment by jasonlin15** [comment]
Interesting PR

**Comment by MaximilianMeiler** [comment]
> Interesting PR

https://tenor.com/view/mimimimimi-cat-sleeping-snoring-cat-sleeping-cat-drooling-cat-gif-23979759

File: Backend/canvas.js [hunk 1]
```diff
@@ -65,10 +65,22 @@ export async function getAssignments(canvas_api_token, course_id) {
...
-      return response.data;
+
+      var res = [];
+      response.data.map((a) => {
+        var dueAt = new Date(a.due_at)
+
+        if (a.due_at && dueAt > Date.now()) {
+          res.push([a.id, a.name, a.due_at, a.points_possible]);
+        }
+      })
```
Comment (kthchew) [inline Backend/canvas.js:81]: `map` generally isn't used this way.

```suggestion
      const res = response.data.flatMap((a) => {
        const dueAt = new Date(a.due_at);

        if (a.due_at && dueAt > Date.now()) {
          return [[a.id, a.name, a.due_at, a.points_possible]];
        } else {
          return [];
      })
```

**Review by kthchew — CHANGES_REQUESTED** [review]
Haven't gotten around to testing changes past the login screen, but here's some of my initial thoughts (I may have more comments after testing)

File: Frontend/src/Login.jsx [hunk 2]
```diff
@@ -20,19 +20,35 @@ export default function Login({setLoginTime, apiKey, setApiKey, setUserData}) {
...
+      setCourses(temp.data.courses)
+    }
 
-  const attemptLogin = async (method) => {
-    setLoginTime(Date.now());
+    if (apiKey) {
+      login()
+    }
 
+  }, [apiKey])
```
[the remark this answers is not in the document]
Comment (MaximilianMeiler) [reply Frontend/src/Login.jsx]: Good catch. Just threw up a jank lil fix for this (hopefully)

File: Backend/canvas.js [hunk 3]
```diff
@@ -93,3 +105,37 @@ export async function getSubmissions(canvas_api_token, course_id, assignment_id,
...
+        "order_direction": "descending"
+      }
+    })
+
+    var newSubmissions = [];
+    var submissions = response.data[0].submissions;
+    submissions.map((temp) => {
+      newSubmissions.push([temp.assignment.id, temp.assignment.name, temp.assignment.due_at, temp.assignment.points_possible, 
+        temp.id, temp.submitted_at, temp.score])
+    })
```
[the remark this answers is not in the document]
Comment (MaximilianMeiler) [reply Backend/canvas.js:135]: Replaced.

File: Backend/canvas.js (hunk 1 above)
Comment (MaximilianMeiler) [reply Backend/canvas.js:81]: Clever solution, replaced.

File: Frontend/src/Login.jsx [hunk 4]
```diff
@@ -20,19 +20,35 @@ export default function Login({setLoginTime, apiKey, setApiKey, setUserData}) {
...
+      setCourses(temp.data.courses)
+    }
 
-  const attemptLogin = async (method) => {
-    setLoginTime(Date.now());
+    if (apiKey && logState == "login") {
+      login()
+    }
+
+  }, [apiKey, logState])
```
Comment (kthchew) [inline Frontend/src/Login.jsx:42]: This still has a problem if there's something in the API key box but the user cancels the registration and goes back to the login page.

Does this work? Seems to work when I tried it.

```diff
diff --git a/Frontend/src/Login.jsx b/Frontend/src/Login.jsx
index b2195b7..25c8544 100644
--- a/Frontend/src/Login.jsx
+++ b/Frontend/src/Login.jsx
@@ -20,35 +20,24 @@ export default function Login({apiKey, setApiKey, setUserData, setUserId, setCou
     // alert("Username: " + username + " " + "Password: " + password);
   };
 
-  useEffect(() => {
-    const login = async () => {
-      const temp = await axios.post(`${API_URL}/loginUser`, {
-        username: username,
-        password: password,
-        apiKey: apiKey
-      })
-  
-      console.log(temp)
-      
-      setUserData(temp.data.userData)  
-      setUserId(temp.data.userId)
-      setCourses(temp.data.courses)
-    }
-
-    if (apiKey && logState == "login") {
-      login()
-    }
-
-  }, [apiKey, logState])
-
 
   const attemptLogin = async (method) => {
     if (method === "login") {
       try {
-        setApiKey(localStorage.getItem("canvasAPIKey"))
-        console.log("logged in")
+        const currentKey = localStorage.getItem("canvasAPIKey");
+        setApiKey(currentKey);
+        const result = await axios.post(`${API_URL}/loginUser`, {
+          username: username,
+          password: password,
+          apiKey: currentKey
+        });
 
+        console.log(result);
         
+        setUserData(result.data.userData);
+        setUserId(result.data.userId);
+        setCourses(result.data.courses);
+        console.log("logged in");
       } catch (e) {
         console.log("login failed");
       }

```

File: Backend/server.js [hunk 5]
```diff
@@ -84,22 +84,11 @@ app.get('/getSubmission', async (req, res) => {
...
-
-
 
 app.post('/loginUser', limiter, async (req, res) => {
   const u = req.body.username;
   const p = req.body.password;
+  const apiKey = req.body.apiKey;
   
   if (!u || !p) {
     return res.status(400).json({message: "Username and password required!"});
```
Comment (kthchew) [inline Backend/server.js:94]: The entire server crashes if a single user tries to login without an API key saved. e.g. if you register an account then clear local storage/use a different device/browser to login later, the server will crash. We should make UI to ask for the API key again, but for now we should probably check if it's actually there or not before using it (like it does for username and password already).

This is also a problem in some other endpoints. The same guard should probably be added on those, too.

**Review by kthchew — CHANGES_REQUESTED** [review]

File: Backend/server.js (hunk 5 above)
Comment (MaximilianMeiler) [reply Backend/server.js:94]: What other routes you think would suffer this same issue?

**Review by kthchew — APPROVED** [review]

File: Backend/server.js (hunk 5 above)
Comment (kthchew) [reply Backend/server.js]: I fixed them in 79a4864

**Pull request merged**